File: service/adapters/ppocr-sidecar.mjs
/**
 * Adopted canonical OCR witness (issue #2, PR #67 ceremony): the official
 * PaddleOCR pipeline as a subprocess-per-worker Python sidecar.
 *
 * Protocol: JSONL over stdin/stdout — {id, path} in, {id, lines, ms} out;
 * the child's FIRST line is a meta record (versions, platform, thread
 * setting, model pins, in-band `useHpip`). Pages travel as tmpfiles, not
 * base64. A child crash rejects every pending request (the page fails
 * closed and the queue requeues it); the next recognize() respawns the
 * child. Reaping is by PROCESS-GROUP kill (children spawn detached in
 * their own group) on dispose, worker exit hooks, and worker signals —
 * necessary because a wrapper launcher like uv spawns python rather than
 * exec'ing it, so killing only the direct child would orphan the engine.
 * If the worker itself dies un-hooked (SIGKILL), the fallback is the
 * child's stdin-EOF exit, which waits out any in-flight predict.
 *
 * Provenance is truthful per host: the descriptor's `ep=` is derived from
 * the child's own meta (hpi on Linux, paddle-default elsewhere) — never a
 * claim the process could not verify. The engine-selection line the C++
 * layer prints (e.g. "Backend::OPENVINO") bypasses Python logging, so this
 * adapter retains the child's recent stderr and exposes the first matching
 * line as LOG-DERIVED evidence (`backend.engineEvidence`), labeled as such.
 *
 * Models are pinned (service/sidecar/model-pins.json): the child verifies
 * every file hash before loading and refuses to serve on mismatch — and
 * `verifyModelPins()` lets the server fail closed at boot, before any job
 * is accepted.
 */
import { spawn } from 'node:child_process';
import { createHash, randomUUID } from 'node:crypto';
import { createInterface } from 'node:readline';
import { mkdtempSync, readFileSync, rmSync, unlinkSync, writeFileSync, existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const SIDECAR_DIR = dirname(fileURLToPath(import.meta.url)).replace(/adapters$/u, 'sidecar');
const SIDECAR_SCRIPT = join(SIDECAR_DIR, 'ppocr_sidecar.py');
const PINS_PATH = join(SIDECAR_DIR, 'model-pins.json');
export const DEFAULT_PYTHON_CMD = ['uv', 'run', '--with', 'paddleocr==3.7.0', '--with', 'paddlepaddle==3.2.1', 'python'];

/**
 * Node-side pin verification for boot-time fail-closed (the child verifies
 * again before loading — refusal lives on both sides of the process
 * boundary, same doctrine as the canonical gate).
 */
export function verifyModelPins(modelsDir, pinsPath = PINS_PATH) {
  const manifest = JSON.parse(readFileSync(pinsPath, 'utf8'));
  const failures = [];
  for (const [repo, pin] of Object.entries(manifest.repos)) {
    const base = join(modelsDir, repo.split('/').pop());
    for (const [rel, expected] of Object.entries(pin.files)) {
      const path = join(base, rel);
      if (!existsSync(path)) { failures.push(`${repo}/${rel}: missing`); continue; }
      const got = createHash('sha256').update(readFileSync(path)).digest('hex');
      if (got !== expected) failures.push(`${repo}/${rel}: expected ${expected.slice(0, 12)}, got ${got.slice(0, 12)}`);
    }
  }
  if (failures.length) {
    throw new Error(`Model pin mismatch — refusing to serve unvalidated weights:\n${failures.join('\n')}`);
  }
  return Object.fromEntries(Object.entries(manifest.repos).map(([repo, pin]) => [repo, pin.revision]));
}

const STDERR_RING = 200;

export function createPpOcrSidecarAdapter(config = {}) {
  const { modelsDir, threads = 1, pythonCmd = DEFAULT_PYTHON_CMD, pinsPath = PINS_PATH, metaTimeoutMs = 300_000, recognizeTimeoutMs = 120_000 } = config;
  if (!modelsDir) {
    throw new Error('ppocr-sidecar requires an explicit modelsDir (SERVICE_SIDECAR_MODELS_DIR) — pinned models are configuration, not magic.');
  }
  const pins = verifyModelPins(modelsDir, pinsPath);
  const tmpRoot = mkdtempSync(join(tmpdir(), 'psvc-sidecar-'));

  let child = null;
  let meta = null;
  let coldInitMs = null;
  let disposed = false;
  const pending = new Map();
  const stderrRing = [];
  let nextId = 1;

  /**
   * Kill an instance's WHOLE process group. The default launcher is a uv
   * wrapper that spawns python rather than exec'ing it, so killing only
   * the direct child leaves a ~2 GB python grandchild alive (it then exits
   * only on stdin EOF — never, if it is wedged mid-predict). Children are
   * spawned detached so pgid === child.pid and the group kill reaps the
   * wrapper AND the interpreter.
   */
  function killInstance(instance) {
    if (!instance) return;
    try { process.kill(-instance.pid, 'SIGKILL'); } catch { /* group already gone */ }
    try { instance.kill('SIGKILL'); } catch { /* already dead */ }
  }
  function killChild() {
    const current = child;
    child = null;
    meta = null;
    // Deliberate kills must reject the remaining in-flight pages here: the
    // instance's own exit handler is guarded on `child === self` and child
    // was just nulled, so it will (correctly) not touch shared state.
    rejectPending('OCR sidecar killed.');
    killInstance(current);
  }
  const onExit = () => { killChild(); rmSync(tmpRoot, { recursive: true, force: true }); };
  process.once('exit', onExit);

  function rejectPending(reason) {
    for (const [, entry] of pending) entry.reject(new Error(reason));
    pending.clear();
  }

  async function ensureChild() {
    if (child && meta) return;
    const started = performanceNow();
    const [cmd, ...args] = pythonCmd;
    // detached: the child leads its own process group, so killInstance can
    // reap wrapper + interpreter together (see above).
    const self = spawn(cmd, [...args, SIDECAR_SCRIPT], {
      env: { ...process.env, SIDECAR_MODELS_DIR: modelsDir, SIDECAR_THREADS: String(threads) },
      stdio: ['pipe', 'pipe', 'pipe'],
      detached: true
    });
    child = self;
    // Fresh evidence per child: a respawned engine must not inherit the
    // previous instance's stderr (stale backend lines would lie).
    stderrRing.length = 0;
    // EPIPE window: a write racing child death must reject the page (via
    // the exit handler), never crash the worker.
    self.stdin.on('error', () => { /* pending rejected by exit handler */ });
    // Every handler below is INSTANCE-scoped: a stale child (e.g. one that
    // timed out on meta and died later) must not null the live child's
    // reference or reject the live child's pages.
    self.on('exit', () => {
      if (child !== self) return;
      child = null;
      meta = null;
      rejectPending('OCR sidecar exited.');
    });
    createInterface({ input: self.stderr }).on('line', (line) => {
      if (child !== self) return;
      stderrRing.push(line);
      if (stderrRing.length > STDERR_RING) stderrRing.shift();
    });
    const lines = createInterface({ input: self.stdout });
    meta = await new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        // The slow starter is dead weight AND a hazard: kill its whole
        // group now, or a ~2 GB engine survives unreferenced.
        if (child === self) { child = null; }
        killInstance(self);
        reject(new Error(`Sidecar meta line not received within ${metaTimeoutMs}ms.`));
      }, metaTimeoutMs);
      lines.on('line', (line) => {
        // Stale-instance output is dropped wholesale. (A late meta line
        // after the timeout would hit an already-rejected promise anyway —
        // resolve() there is a no-op.)
        if (child !== self) return;
        let message;
        try { message = JSON.parse(line); } catch { return; }
        if (message.kind === 'meta') {
          clearTimeout(timer);
          resolve(message);
          return;
        }
        if (message.kind === 'fatal') {
          clearTimeout(timer);
          child = null;
          killInstance(self);
          reject(new Error(`Sidecar refused to start: ${message.error}`));
          return;
        }
        const entry = pending.get(message.id);
        if (!entry) return;
        pending.delete(message.id);
        if (message.error) entry.reject(new Error(`Sidecar page error: ${message.error}`));
        else entry.resolve(message);
      });
      self.on('error', (error) => { clearTimeout(timer); if (child === self) { child = null; } reject(error); });
    });
    coldInitMs = Math.round(performanceNow() - started);
  }

  function performanceNow() { return Number(process.hrtime.bigint() / 1_000_000n); }

  function descriptorFor(currentMeta) {
    // ep is derived from the child's own testimony, never asserted: hpi is
    // only claimable where the pipeline itself reports it engaged.
    const ep = currentMeta.useHpip === true ? 'hpi' : 'paddle-default';
    const version = currentMeta.versions?.paddleocr ?? 'unknown';
    return { ep, version, descriptor: `ppocrv6-small-sidecar@${version}#ep=${ep};threads=${threads}` };
  }

  return {
    name: 'ppocrv6-small-sidecar',
    get version() { return meta?.versions?.paddleocr ?? 'unpinned'; },
    canonical: true,
    get descriptor() { return meta ? descriptorFor(meta).descriptor : `ppocrv6-small-sidecar@unstarted#ep=unknown;threads=${threads}`; },
    get backend() {
      if (!meta) return null;
      const { ep, version } = descriptorFor(meta);
      return {
        adapter: 'ppocrv6-small-sidecar',
        version,
        variant: 'small',
        executionProvider: ep,
        numThreads: threads,
        platform: meta.platform,
        hpiRequested: meta.hpiRequested,
        useHpip: meta.useHpip,
        modelPins: pins,
        // The C++ engine line bypasses Python logging entirely; this is the
        // best available evidence and is labeled for what it is.
        engineEvidence: {
          source: 'log-derived (child stderr)',
          // Prefer the latest concrete C++ engine selection. A later generic
          // "backend config" line may carry thread settings but cannot prove
          // OpenVINO rather than Paddle served the request.
          line: stderrRing.filter((line) => /Backend::/u.test(line)).at(-1)
            ?? stderrRing.filter((line) => /backend config/u.test(line)).at(-1)
            ?? null
        }
      };
    },
    get sidecarMeta() { return meta; },
    async warmup() {
      await ensureChild();
    },
    coldInitMs: () => coldInitMs,
    async dispose() {
      disposed = true;
      rejectPending('OCR sidecar disposed.');
      killChild();
      process.removeListener('exit', onExit);
      rmSync(tmpRoot, { recursive: true, force: true });
    },
    async recognize(page) {
      if (disposed) throw new Error('OCR sidecar disposed.');
      await ensureChild();
      if (disposed) throw new Error('OCR sidecar disposed.');
      const id = nextId++;
      const path = join(tmpRoot, `${randomUUID()}.png`);
      writeFileSync(path, Buffer.from(page.data));
      try {
        const response = await new Promise((resolve, reject) => {
          // A torn or dropped response line (the C++ layer writes to the raw
          // fds; unparseable lines are skipped) must not park this promise
          // forever — that is silent pool starvation. On deadline: fail the
          // page closed (queue requeues) and kill the child, whose protocol
          // state is now unknowable.
          const deadline = setTimeout(() => {
            pending.delete(id);
            killChild();
            reject(new Error(`Sidecar page timed out after ${recognizeTimeoutMs}ms; child killed, page fails closed.`));
          }, recognizeTimeoutMs);
          pending.set(id, {
            resolve: (value) => { clearTimeout(deadline); resolve(value); },
            reject: (error) => { clearTimeout(deadline); reject(error); }
          });
          try {
            child.stdin.write(`${JSON.stringify({ id, path })}\n`);
          } catch (error) {
            clearTimeout(deadline);
            pending.delete(id);
            reject(error);
          }
        });
        const observations = (response.lines ?? [])
          .filter((line) => typeof line.text === 'string' && line.text.trim())
          .map((line, index) => {
            const score = line.score;
            if (!Number.isFinite(score) || score < 0 || score > 1) {
              throw new Error(`Sidecar line ${index} has invalid confidence ${score}.`);
            }
            const polygon = Array.isArray(line.poly) ? line.poly.map((point) => [point[0], point[1]]) : null;
            if (!polygon || polygon.length < 3) {
              throw new Error(`Sidecar line ${index} is missing its polygon.`);
            }
            const xs = polygon.map((point) => point[0]);
            const ys = polygon.map((point) => point[1]);
            return {
              id: `ppocr-sidecar:${page.pageNumber}:${index}`,
              pageNumber: page.pageNumber,
              text: line.text,
              box: [Math.min(...xs), Math.min(...ys), Math.max(...xs), Math.max(...ys)],
              polygon,
              confidence: score,
              model: 'PP-OCRv6_small'
            };
          });
        return { pageNumber: page.pageNumber, observations, backend: descriptorFor(meta).ep };
      } finally {
        try { unlinkSync(path); } catch { /* tmpfile already gone */ }
      }
    }
  };
}

File: service/sidecar/ppocr_sidecar.py
"""PP-OCRv6 sidecar — the adopted canonical OCR witness (issue #2).

One Python child per Node page-worker, speaking JSONL over stdin/stdout:
  request:  {"id": 7, "path": "/tmp/....png"}
  response: {"id": 7, "lines": [{"text","poly","score"}], "ms": 123.4}
  errors:   {"id": 7, "error": "..."}       (page fails closed upstream)

The FIRST stdout line is the meta record — versions, platform, thread
setting, os.cpu_count, model pins, and backend truth:
  - useHpip: introspected in-band from the pipeline object (the same
    attributes the adoption ceremony verified).
  - The engine-selection line (Backend::OPENVINO ...) is emitted by the
    C++ layer to stderr and CANNOT be captured in-band here; the Node
    adapter captures this process's stderr and labels that evidence
    log-derived. Meta carries hpiRequested/useHpip only — never a claim
    about the engine that this process cannot itself observe.

Models are LOCAL AND PINNED (fetch_models.py): this process verifies the
pin manifest hashes before loading and refuses to serve on any mismatch —
a host can never run weights outside the validated lineage.

enable_hpi is requested only on Linux (upstream HPI support); elsewhere the
pipeline runs paddle-default and meta says so. Provenance is truthful per
host by construction.
"""

import hashlib
import json
import os
import platform
import sys
import time
from pathlib import Path

MANIFEST = Path(__file__).parent / "model-pins.json"


def fail(msg: str) -> None:
    print(json.dumps({"kind": "fatal", "error": msg}), flush=True)
    sys.exit(1)


def verify_pins(models_dir: Path) -> dict:
    if not MANIFEST.exists():
        fail("model-pins.json missing — run fetch_models.py --record and commit it")
    manifest = json.loads(MANIFEST.read_text())
    pins = {}
    for repo, pin in manifest["repos"].items():
        target = models_dir / repo.split("/")[-1]
        for rel, expected in pin["files"].items():
            path = target / rel
            if not path.is_file():
                fail(f"pinned model file missing: {path}")
            digest = hashlib.sha256()
            with path.open("rb") as handle:
                for chunk in iter(lambda: handle.read(1 << 20), b""):
                    digest.update(chunk)
            if digest.hexdigest() != expected:
                fail(f"pin mismatch for {repo}/{rel}: refusing to serve unvalidated weights")
        pins[repo] = pin["revision"]
    return pins


def _os_thread_count():
    try:
        return len(os.listdir("/proc/self/task"))
    except OSError:
        return None


def main() -> None:
    models_dir = Path(os.environ["SIDECAR_MODELS_DIR"])
    threads = int(os.environ.get("SIDECAR_THREADS", "1"))
    check_only = "--check" in sys.argv

    # SIDECAR_THREADS is the ONLY thread knob. `cpu_threads=` below never
    # reaches the OpenVINO/ONNX HPI runners: PaddleX sizes those pools from
    # PADDLE_PDX_CPU_NUM_THREADS (default 10) at pipeline construction
    # (paddlex/inference/models/runners/hpi/config.py). Left unset, four
    # sidecars on four cores ran 40 inference threads (issue #126).
    os.environ["PADDLE_PDX_CPU_NUM_THREADS"] = str(threads)

    t0 = time.monotonic()
    pins = verify_pins(models_dir)

    import importlib.metadata

    from paddleocr import PaddleOCR

    # SIDECAR_DISABLE_HPI=1 forces paddle-default on Linux — the EP-control
    # knob the era rule requires (same host, same models, HPI off). Meta and
    # the descriptor stay truthful either way: useHpip is introspected, not
    # asserted.
    hpi_requested = platform.system() == "Linux" and os.environ.get("SIDECAR_DISABLE_HPI") != "1"
    kwargs = dict(
        text_detection_model_name="PP-OCRv6_small_det",
        text_detection_model_dir=str(models_dir / "PP-OCRv6_small_det"),
        text_recognition_model_name="PP-OCRv6_small_rec",
        text_recognition_model_dir=str(models_dir / "PP-OCRv6_small_rec"),
        use_doc_orientation_classify=False,
        use_doc_unwarping=False,
        use_textline_orientation=False,
        device="cpu",
    )
    if hpi_requested:
        kwargs["enable_hpi"] = True
    try:
        ocr = PaddleOCR(**kwargs, cpu_threads=threads)
        thread_kwarg_applied = True
    except TypeError:
        ocr = PaddleOCR(**kwargs)
        thread_kwarg_applied = False
    init_s = time.monotonic() - t0

    # In-band backend truth: the same attributes the adoption ceremony
    # introspected. Absence is reported as null, never guessed.
    def probe_use_hpip():
        for holder in (getattr(ocr, "paddlex_pipeline", None), ocr):
            for attr in ("use_hpip", "_use_hpip"):
                value = getattr(holder, attr, None)
                if isinstance(value, bool):
                    return value
                inner = getattr(holder, "_pipeline", None)
                value = getattr(inner, attr, None) if inner is not None else None
                if isinstance(value, bool):
                    return value
        return None

    meta = {
        "kind": "meta",
        "versions": {
            pkg: importlib.metadata.version(pkg)
            for pkg in ("paddleocr", "paddlepaddle")
        },
        "platform": platform.system(),
        "machine": platform.machine(),
        "osCpuCount": os.cpu_count(),
        "threads": threads,
        "threadKwargApplied": thread_kwarg_applied,
        # In-band attestation for the thread setting (issue #126 step 1):
        # the env PaddleX read, and this process's live OS thread count
        # after engine construction (Linux only; null elsewhere).
        "pdxCpuNumThreadsEnv": os.environ.get("PADDLE_PDX_CPU_NUM_THREADS"),
        "osThreadCount": _os_thread_count(),
        "hpiRequested": hpi_requested,
        "useHpip": probe_use_hpip(),
        "modelPins": pins,
        "initS": round(init_s, 2),
    }
    print(json.dumps(meta), flush=True)
    if check_only:
        return

    import cv2

    for raw in sys.stdin:
        raw = raw.strip()
        if not raw:
            continue
        try:
            request = json.loads(raw)
        except json.JSONDecodeError:
            print(json.dumps({"kind": "fatal", "error": "bad request framing"}), flush=True)
            sys.exit(1)
        rid = request.get("id")
        try:
            image = cv2.imread(request["path"], cv2.IMREAD_COLOR)
            if image is None:
                raise ValueError(f"unreadable image: {request['path']}")
            t = time.monotonic()
            result = ocr.predict(image)
            ms = (time.monotonic() - t) * 1000
            lines = []
            for res in result:
                data = res if isinstance(res, dict) else res.json.get("res", res.json)
                texts = data.get("rec_texts", [])
                scores = data.get("rec_scores", [])
                polys = data.get("rec_polys", data.get("dt_polys", []))
                for i, text in enumerate(texts):
                    poly = polys[i] if i < len(polys) else None
                    lines.append(
                        {
                            "text": str(text),
                            "poly": [[float(p[0]), float(p[1])] for p in poly]
                            if poly is not None
                            else None,
                            "score": float(scores[i]) if i < len(scores) else None,
                        }
                    )
            print(json.dumps({"id": rid, "lines": lines, "ms": round(ms, 1)}), flush=True)
        except Exception as error:  # per-request containment: page fails closed upstream
            print(json.dumps({"id": rid, "error": f"{type(error).__name__}: {error}"}), flush=True)


if __name__ == "__main__":
    main()
